---
title: "w3 practical"
output: html_document
date: "2024-03-07"
---


```{r}
suppressPackageStartupMessages({
  library(AnnotationHub) # to retrieved to genomes / annotations
  library(Rsubread) # alignment
  library(rtracklayer) # to import/export files
  library(Biostrings) # handle sequences (DNA/protein/RNA)
  library(Rfastp) # read adapter trimming, QC of reads
  library(epiwraps) # visualization
})
ah <- AnnotationHub()
# if on renku, use ah <- AnnotationHub(cache="ahcache")
```

We will perform the primary analysis of an example dataset, at each step demonstrating the use of both a very standard pipeline, and a full R-based equivalent. Pay attention to which chunks are R or bash code (see header)

# Downloading the raw reads

We download an example dataset which is a ChIP-seq experiment mapping the bindings
of the factor Myc across the Kc167 Drosophila cell line.
(Note that it's actually quite a bad experiment)

We download the data from ENCODE:

```{r, eval=FALSE}
options(timeout=3600) # we need to increase the download timeout since the file is large and cnx slow
dir.create("raw")
download.file("https://www.encodeproject.org/files/ENCFF014MOQ/@@download/ENCFF014MOQ.fastq.gz", dest="raw/Myc.fastq.gz", mode = "wb")
download.file("https://www.encodeproject.org/files/ENCFF179TMO/@@download/ENCFF179TMO.fastq.gz", dest="raw/input.fastq.gz", mode="wb")
```


# Reads QC and trimming

## Using fastqc 

For this we will use shell commands:

```{bash engine.opts='-l', eval=FALSE}
# note that depending on how your computer/server is set up, you might need to 
# specify where to find the software, e.g.:
export PATH="/opt/miniconda3/bin:$PATH"
mkdir -p raw_fastqc
fastqc -o raw_fastqc --threads 4 raw/*.fastq.gz
```

This will create an html report with QC plots for each file. We could optionally aggregate those using multiQC:

```{bash engine.opts='-l', eval=FALSE}
export PATH="/opt/miniconda3/bin:$PATH"
multiqc -o raw_fastqc raw_fastqc
```


### Trimming

```{bash, eval=FALSE}
export PATH="/opt/miniconda3/bin:$PATH"

mkdir -p trimmed
# this would be for a single sample:
trimmomatic SE -threads 4 -summary trimmed/Myc.stats -phred33 raw/Myc.fastq.gz trimmed/Myc.fastq.gz \
ILLUMINACLIP:TruSeq3-SE:2:15:4:4:true LEADING:20 TRAILING:20 SLIDINGWINDOW:4:15 MINLEN:25

trimmomatic SE -threads 4 -summary trimmed/input.stats -phred33 raw/input.fastq.gz trimmed/input.fastq.gz ILLUMINACLIP:TruSeq3-SE:2:15:4:4:true LEADING:20 TRAILING:20 SLIDINGWINDOW:4:15 MINLEN:25
```

## Using R and Rfastp

```{r}
dir.create("rfastp.trimmed")

fastq_files <- c(Myc="raw/Myc.fastq.gz", input="raw/input.fastq.gz")

qc <- lapply(fastq_files, FUN=function(x){
  Rfastp::rfastp(x, thread=4, overrepresentationAnalysis=TRUE,
                 outputFastq=file.path("rfastp.trimmed/",gsub("\\.fastq\\.gz$","",basename(x))))
})
```

This will create an html report with before/after QC plots, as well as a trimmed
and filtered fastq file. Furthermore, the R output object (equivalent to the .json 
stats also saved by the program) can be used to generate the QC plots inside a
markdown, for example:

```{r}
Rfastp::curvePlot(qc$Myc, curve="content_curves")
```


# Alignment

## Using Rsubread

### Building a genome index for mapping

This we have to do only once for a genome, and can then re-use across projects

```{r, eval=FALSE}
# we get the genome sequence from AnnotationHub
genome <- ah[["AH49674"]]
# we create a new directory that will contain the genome index
dir.create("BDGP6_genome")
# we write the genome sequence in fasta format
export(import.2bit(genome), "BDGP6_genome/genome.fasta.gz", compress=TRUE)
# we build a Rsubread index
Rsubread::buildindex("BDGP6_genome/rsubread", reference="BDGP6_genome/genome.fasta.gz")
```

### Alignment

```{r}
# to know how many CPUs are available:
parallel::detectCores()


dir.create("aligned")
align.stats <- Rsubread::align(index="BDGP6_genome/rsubread", type="dna",
                               readfile1=c("rfastp.trimmed/Myc_R1.fastq.gz", 
                                           "rfastp.trimmed/input_R1.fastq.gz"),
                               output_file=c("aligned/Myc.bam","aligned/input.bam"),
                               nthreads=6, sortReadsByCoordinates=TRUE)
align.stats

```


## Using bowtie2

### Building a bowtie2 index

This we would need to run only once for this genome:

```{bash engine.opts='-l'}
# bowtie2-build [options] path/to/genome.fasta path/to/outputIndex
bowtie2-build --threads 4 BDGP6_genome/genome.fasta.gz BDGP6_genome/bowtie2
```

### Alignment

```{bash engine.opts='-l'}
mkdir -p aligned
(bowtie2 -p 4 -x BDGP6_genome/bowtie2 -U trimmed/Myc.fastq.gz) 2> aligned/Myc.bowtie2 |\
  samtools view -bS - | samtools sort -@4 -m 2G - > aligned/Myc.bam
samtools index aligned/Myc.bam
(bowtie2 -p 4 -x BDGP6_genome/bowtie2 -U trimmed/input.fastq.gz) 2> aligned/input.bowtie2 |\
  samtools view -bS - | samtools sort -@4 -m 2G - > aligned/input.bam
samtools index aligned/input.bam
```

# Peak calling

## Using MACS2

```{bash engine.opts='-l'}
export PATH="/opt/miniconda3/bin:$PATH"
mkdir -p peaks
macs3 callpeak --outdir peaks -n Myc --gsize dm -t aligned/Myc.bam -c aligned/input.bam
# macs2 works in the same way
```

We can then import the peaks in R:

```{r}
peaks <- rtracklayer::import("peaks/Myc_peaks.narrowPeak")
peaks
```


## Using R

This alternative is currently still very beta when it comes to using the input control, so here we will do without it. For the purpose of this course, this is offered as an alternative for windows users who can't install MACS2/3:

```{r}
peaks <- callPeaks("aligned/Myc.bam", fragLength=50L)
# if we want to save it as a bed file:
rtracklayer::export.bed(peaks, "peaks/peaks.bed")
```

## Having a loog at a peak

Now it turns out that this is a particularly bad datasets, and many peaks (no matter the caller used) aren't very convincing when we compare to the input signal, for example:

```{r}
peaks[1]
plotSignalTracks(files = c(Myc="aligned/Myc.bam", Input="aligned/input.bam"), region=peaks[1], extend=1000)
```

Here's instead a good example:

```{r}
# peak 141 from the epiwraps peak, or peak 3246 from the macs peaks, or:
myGreatPeak <- "2L:19582199-19582565"
plotSignalTracks(c(Myc="aligned/Myc.bam", Input="aligned/input.bam"), region=myGreatPeak,
                 extend=1000)

# we set the y-axis scales to the same range:
plotSignalTracks(c(Myc="aligned/Myc.bam", Input="aligned/input.bam"), region=myGreatPeak,
                 extend=1000, tracks.params=list(ylim=c(0,50)))
```

