---
title: "Exploration of processed ChIP-seq data"
author: "Pierre-Luc"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(rtracklayer)
  library(epiwraps)
  library(GenomicRanges)
})
ah <- AnnotationHub()

# not for today:
# example to check metadata columns of annotation objects:
q <- query(ah, c("GRCm38", "EnsDb"))
q
qmetadata <- as.data.frame(mcols(q))
View(qmetadata)
```


# Downloading the processed data

We will follow-up on what we did last week, starting with already aligned reads and peaks.
Since the data used last week was of bad quality, here we'll use a different dataset (ChIP for CTCF in Drosophila S2 cells).
To improve speed and so on, we'll work with a subset of the data on the chromosome 2L.

You can download this data like this:

```{r, eval=FALSE}
download.file("https://ethz-ins.org/content/w04_practical.zip", "w04_practical.zip")
unzip("w04_practical.zip")
```



# Visualizing a single region

```{r}
peaks <- rtracklayer::import("peaks/ctcf.narrowPeak")
# when we download from ENCODE, peaks lack the narrowPeak extension but are still
# in that format, so we might need to specify the format:
peaks <- rtracklayer::import("peaks/ctcf.narrowPeak", format="narrowPeak")
head(peaks)
peaks[1]
plotSignalTracks( files=list(CTCF="aligned/ctcf.bam", Input="aligned/input.bam", peaks=peaks), region=peaks[1], extend=2000)
# put them on the same scale:
plotSignalTracks( files=list(CTCF="aligned/ctcf.bam", Input="aligned/input.bam", peaks=peaks), region=peaks[1], extend=2000, tracks.params = list(ylim=c(0,55)))

plotSignalTracks( files=list(CTCF="aligned/ctcf.bam", Input="aligned/input.bam", peaks=peaks), region=peaks[1], extend=1000, type="alignment")
```


# Generating coverage tracks

Again here I'm providing two ways of doing this, first the non-R version that's the most widely used, and then a R-based version, both of which should be equivalent.

## Using DeepTools


```{bash engine.opts='-l', eval=FALSE}
mkdir -p tracks
for bamfile in aligned/*.bam; do
  bamCoverage -p 4 --binSize 20 --normalizeUsing CPM -b $bamfile -o tracks/`basename $bamfile .bam`.bw
done
```


## Using R

In this example I'll showcase the binning by generating two tracks (i.e. bigwigs) for the IP:

```{r}
dir.create("tracks")
epiwraps::bam2bw("aligned/ctcf.bam", "tracks/ctcf_full.bw", binWidth=1, extend=100, scaling=TRUE)
epiwraps::bam2bw("aligned/ctcf.bam", "tracks/ctcf.bw", binWidth=20, extend=100, scaling=TRUE)
epiwraps::bam2bw("aligned/input.bam", "tracks/input.bw", binWidth=20, extend=100, scaling=TRUE)
```




# Looking at an example peak

Just to compare binning vs not:

```{r}
tracks <- list(ctcf_binWidth1="tracks/ctcf_full.bw", ctcf_binWidth20="tracks/ctcf.bw", peaks=peaks)
plotSignalTracks( files=tracks, region=peaks[1], extend=200)
```

```{r}
tracks <- list(ctcf="tracks/ctcf_full.bw", Input="tracks/input.bw", peaks=peaks)
plotSignalTracks( files=tracks, region=peaks[1], extend=1000)
plotSignalTracks( files=tracks, region=peaks[1], extend=1000, tracks.params = list(ylim=c(0,180)))
```

```{r}
ensdb <- ah[["AH95713"]] # Drosophila EnsDb
plotSignalTracks( files=tracks, region=peaks[1], extend=1000,
                  tracks.params = list(ylim=c(0,180)),
                  ensdb=ensdb, transcripts = "full")
```

The TF Myc, with which we worked last week, is instead typically found at promoters: 

```{r}
myc_peaks <- rtracklayer::import("peaks/myc.narrowPeak")
myc_tracks <- list(Myc="tracks/myc.bw", peaks=myc_peaks)
plotSignalTracks( files=myc_tracks, region=myc_peaks[1],
                  ensdb=ensdb, transcripts = "full", extend=2000)
# note that instead of querying a region, we can also simply query a gene name
# (assuming we provided an ensdb object)
plotSignalTracks( files=myc_tracks, region="numb", ensdb=ensdb, transcripts = "full")
```


# Looking at multiple peaks

```{r}
tracks <- list(ctcf="tracks/ctcf.bw", Input="tracks/input.bw")
sm <- epiwraps::signal2Matrix(filepaths = tracks, regions = peaks)
plotEnrichedHeatmaps(sm)

# (requires epiwraps version >=0.99.83 -> update with BiocManager::install("ETHZ-INS/epiwraps") )
# if we want to look at the most enriched half of the peaks:
meanEnrScore <- rowMeans(as.matrix(assay(sm, "enriched_score")))
hist(log1p(meanEnrScore))
# subset to the peaks with a mean enrichment score
sm_high <- sm[which(meanEnrScore>median(meanEnrScore)),]
plotEnrichedHeatmaps(sm_high)
# get back those peaks:
rowRanges(sm_high)

# sm
# plotEnrichedHeatmaps(sm_high[,1])
```

## Computing signal relative to the input

```{r}
epiwraps::bam2bw("aligned/ctcf.bam", bgbam="aligned/input.bam", 
                 output_bw="tracks/ctcf_relative.bw", binWidth=20, extend=100, scaling=TRUE)
```

To visualize this, we'll compare the original signal with the relative one:

```{r}
peakSignals <- signal2Matrix(c(
  ctcf_raw="tracks/ctcf.bw", ctcf_relative="tracks/ctcf_relative.bw", 
                               input="tracks/input.bw"),
                             regions=peaks, extend=2000, w=20)
plotEnrichedHeatmaps(peakSignals[,c(1,3)], scale_title="raw") + 
  plotEnrichedHeatmaps(peakSignals[,2], scale_title="normalized\nto input")
```
In contrast, Myc looks pretty bad:

```{r}
sm_myc <- signal2Matrix("tracks/myc.bw", regions=myc_peaks)
plotEnrichedHeatmaps(sm_myc)
```




# Manipulating peaks

```{r}
peaks
```


## How many peaks overlap promoters

```{r}
txs <- transcripts(ensdb)
head(txs)
proms <- promoters(txs, upstream=1000, downstream=500)
head(proms)
peaksOverlapsPromoters <- overlapsAny(peaks, proms)
head(peaksOverlapsPromoters)
table(peaksOverlapsPromoters)
```

Overlaps are not symmetrical:

```{r}
sum(peaksOverlapsPromoters)
promotersOverlapPeaks <- overlapsAny(proms, peaks)
sum(promotersOverlapPeaks)
```


## Basic peak annotation

```{r}
annotatedPeaks <- epiwraps::annotateRegions(peaks, ensdb)
head(annotatedPeaks)
```

```{r}
table(annotatedPeaks$class)
hist(annotatedPeaks$distance2nearestTSS, breaks = 100)
```

If we want to extract the subset of peaks that are within promoters:

```{r}
peaks$overlapsPromoter <- overlapsAny(peaks, proms)\
head(peaks)
peaksInPromoters <- peaks[peaks$overlapsPromoter]
length(peaks)
length(peaksInPromoters)
```

