---
title: "Assignment week 7"
author: "Selina Joy Frei"
date: "2024-04-21"
output: html_document
---

In the same dataset of ATAC on chr19, plot 1) the insertion (i.e. ‘cuts’) profile of nucleosome-free fragments and 2) the centers of nucleosome-containing fragments, around the high-confidence motifs of two factors.

You can choose your own factors of interest, or for instance use KLF4 and MAZ.

Expected form of the answer: 2 figures (one for each factor), each containing the two signals around the motifs
```{r}
setwd("/mnt/gapp/work/selina/Epigenomics course/Week 7")
suppressPackageStartupMessages({
  library(epiwraps)
  library(AnnotationHub)
  library(MotifDb)
  library(memes)
  library(universalmotif)
  library(ensembldb)
  library(ggplot2)
})

ah <- AnnotationHub(localHub=TRUE)
```

Download dataset
```{r}
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam", "atac.chr19.bam")
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam.bai", "atac.chr19.bam.bai")
```


## Preparing tracks for nucleosome-free fragments cuts + nucleosome-containing fragments

```{r}
bam <- "atac.chr19.bam"

# create a track using nucloesome-free fragments 
epiwraps::bam2bw(bam, output_bw = "NF_cuts.bw", paired=TRUE, binWidth=1L, type="ends", extend=2L, minFragLength=30, 
       maxFragLength=120, shift=c(4L,-5L), forceSeqlevelsStyle = "Ensembl")

# create a track using only the (10bp) centers of mono-nucleosome fragments
bam2bw(bam, output_bw = "mono_centers.bw", paired=TRUE, binWidth=5L, minFragLength=140, shift=c(4L,-5L), 
       maxFragLength=220, type="center", extend=10L, forceSeqlevelsStyle = "Ensembl")
```

# Obtaining the sites with a KLF4 motif

```{r}
# get KLF4 motif
motif <- MotifDb::query(MotifDb, c("KLF4","Mus"))[[1]]
motif2 <- convert_motifs(motif, class="TFBSTools-PFMatrix")
genome <- ah[["AH68356"]]
# get the sequence for chr19:
chr19 <- import(genome)["19"]

# find motif matches across chr19
moi_KLF4 <- motifmatchr::matchMotifs(motif2, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi_KLF4 <- as(setNames(moi_KLF4,names(chr19)), "GRanges")
```

# Obtaining the sites with a MAZ motif

```{r}
# get MAZ motif
motif_MAZ <- MotifDb::query(MotifDb, c("MAZ","Mus"))[[1]]
motif_MAZ2 <- convert_motifs(motif_MAZ, class="TFBSTools-PFMatrix")

# find motif matches across chr19
moi_MAZ <- motifmatchr::matchMotifs(motif_MAZ2, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi_MAZ <- as(setNames(moi_MAZ,names(chr19)), "GRanges")
```


Plotting signal around KLF4 motif

```{r, fig.width=8, fig.height=4}
# we prepare the list of tracks
tracks <- c("NF cuts"="NF_cuts.bw",  "Mono centers"="mono_centers.bw")

# extract signals around the KLF4 motif occurences
sm_KLF4 <- signal2Matrix(tracks, moi_KLF4, w=5, extend=300)

# background normalization to make sure that we have comparable signals
# then we apply the normalization factors:
nf <- getNormFactors(tracks, useSeqLevels="19", nwind=5000L)
sm_KLF4 <- renormalizeSignalMatrices(sm_KLF4, scaleFactors = nf)
plotEnrichedHeatmaps(sm_KLF4, trim=0.95, colors = c("white","darkred"), minRowVal = 10)

# we subset to the top 1000 regions:
sm.top.KLF4 <- head(sm_KLF4[order(rowMeans(as.matrix(assays(sm_KLF4)$enriched_score))),], 1000)
plotEnrichedHeatmaps(sm.top.KLF4,trim=0.95,   colors = c("white","darkred"))

# As expected: at the motif itself we have low signal since there the TF binds and the enzyme cannot cut. 
```

Plotting signal around MAZ motif
```{r, fig.width=8, fig.height=4}
# extract signals around the MAZ motif occurences
sm_MAZ <- signal2Matrix(tracks, moi_MAZ, w=5, extend=300)

# background normalization to make sure that we have comparable signals
# then we apply the normalization factors:
sm_MAZ <- renormalizeSignalMatrices(sm_MAZ, scaleFactors = nf)
plotEnrichedHeatmaps(sm_MAZ, trim=0.95, colors = c("white","darkred"), minRowVal = 10, use_raster = FALSE)

# we subset to the top 1000 regions:
#sm.top.MAZ <- head(sm_MAZ[order(rowMeans(as.matrix(assays(sm_MAZ)$enriched_score))),], 2000)
# plotEnrichedHeatmaps(sm.top.MAZ, trim=0.95,  colors = c("white","darkred"))
```
